recommender/services/synthetic_dataset/users.py

```diff
@@ -0,0 +1,141 @@
+# pylint: disable=invalid-name, no-member, missing-module-docstring
+
+import json
+import os
+from typing import List, Dict, Tuple, Optional
+
+import numpy as np
+import pandas as pd
+from inflection import underscore, pluralize
+
+from definitions import ROOT_DIR
+from recommender.engine.preprocessing import (
+    USERS,
+    create_transformer,
+    precalculate_tensors,
+    save_transformer,
+)
+from recommender.models import User, Service, Category, ScientificDomain
+
+SYNTHETIC_USERS = "synthetic_users"
+CLUSTER_NAMES = [
+    "humanities",
+    "earth",
+    "medicine",
+    "biology",
+    "physics",
+    "computer_science",
+    "business",
+]
+CATEGORY_CLUSTERS_PATH = os.path.join(ROOT_DIR, "resources", "category_clusters.json")
+SCIENTIFIC_DOMAIN_CLUSTERS_PATH = os.path.join(
+    ROOT_DIR, "resources", "scientific_domain_clusters.json"
+)
+
+
+def _sample_niche(
+    cluster: str, clusters_json: Dict[str, List], sample_range: Tuple[int, int] = (0, 3)
+) -> List[str]:
+    """Sample either categories or scientific_domains
+    that are associated with given cluster."""
+    matching_niches = list(
+        map(lambda x: x[0], filter(lambda x: cluster in x[1], clusters_json.items()))
+    )
+    n = min(
+        np.random.randint(sample_range[0], sample_range[1] + 1), len(matching_niches)
+    )
+    return np.random.choice(matching_niches, n, replace=False).tolist()
+
+
+def _filter_relevant(clazz) -> List[str]:
+    """Returns either categories or scientific_domains that
+    have more than 1 service associated with them"""
+    assert clazz in [Category, ScientificDomain]
+    histogram = {}
+
+    for x in clazz.objects:
+        count = len(
+            Service.objects(**{f"{pluralize(underscore(clazz.__name__))}__in": [x]})
+        )
+        if histogram.get(x.name):
+            histogram[x.name] += count
+        else:
+            histogram[x.name] = count
+
+    histogram_df = pd.DataFrame(
+        list(histogram.items()), columns=["class", "count"]
+    ).set_index("class")
+    return histogram_df[histogram_df["count"] >= 1].index.values.tolist()
+
+
+def _synthesize_user(
+    cluster: str,
+    category_clusters: Dict[str, List],
+    scientific_domain_clusters: Dict[str, List],
+    niche_range: Tuple[int, int] = (0, 3),
+) -> User:
+    """Synthesizes a single user using the category/scientific_domain
+    mapping to each cluster"""
+
+    sampled_categories = _sample_niche(cluster, category_clusters, niche_range)
+    sampled_scientific_domains = _sample_niche(
+        cluster, scientific_domain_clusters, niche_range
+    )
+
+    user = User(
+        id=max(User.objects.distinct("id") + [-1]) + 1,
+        categories=Category.objects(name__in=sampled_categories),
```

Comment:
Why not by id rather than name?

Reply:
The sampled categories are sampled from a .json file which defines clustering by name